MagneticField/GeomBuilder/plugins/dd4hep/DD4hep_VolumeBasedMagneticFieldESProducerFromDB.cc

```diff
@@ -0,0 +1,224 @@
+  setWhatProduced(this, label)
+      .setMayConsume(
+          mayConsumeBlobToken_,
+          [](auto const& iGet, edm::ESTransientHandle<MagFieldConfig> iConfig) {
+            if (iConfig->version == "parametrizedMagneticField") {
+              return iGet.nothing();
+            }
+            return iGet("", std::to_string(iConfig->geometryVersion));
+          },
+          edm::ESProductTag<MagFieldConfig, IdealMagneticFieldRecord>(myConfigTag))
+      .setConsumes(chosenConfigToken_, myConfigTag);  //Use same tag as the choice
+}
+
+std::shared_ptr<MagFieldConfig const> DD4hep_VolumeBasedMagneticFieldESProducerFromDB::chooseConfigAtRuntime(
+    IdealMagneticFieldRecord const& iRcd) {
+  edm::ESHandle<MagFieldConfig> config = iRcd.getHandle(mayGetConfigToken_);
+
+  //just forward what we just got but do not take ownership
+  return std::shared_ptr<MagFieldConfig const>(config.product(), [](auto*) {});
+}
+
+std::shared_ptr<MagFieldConfig const> DD4hep_VolumeBasedMagneticFieldESProducerFromDB::chooseConfigViaParameter(
+    const IdealMagneticFieldRecord& iRecord) {
+  auto config = iRecord.getHandle(knownFromParamConfigToken_);
+
+  //just forward what we just got but do not take ownership
+  return std::shared_ptr<MagFieldConfig const>(config.product(), [](auto*) {});
+}
+
+// ------------ method called to produce the data  ------------
+std::unique_ptr<MagneticField> DD4hep_VolumeBasedMagneticFieldESProducerFromDB::produce(
+    const IdealMagneticFieldRecord& iRecord) {
+  auto const& conf = iRecord.getTransientHandle(chosenConfigToken_);
+
+  std::unique_ptr<MagneticField> paramField =
+      ParametrizedMagneticFieldFactory::get(conf->slaveFieldVersion, conf->slaveFieldParameters);
+
+  edm::LogInfo("MagneticFieldDB") << "Version: " << conf->version << " geometryVersion: " << conf->geometryVersion
+                                  << " slaveFieldVersion: " << conf->slaveFieldVersion;
+
+  if (conf->version == "parametrizedMagneticField") {
+    // The map consist of only the parametrization in this case
+    return paramField;
+  }
+
+  // Full VolumeBased map + parametrization
+  MagGeoBuilder builder(conf->version, conf->geometryVersion, debug_);
+
+  // Set scaling factors
+  if (!conf->keys.empty()) {
+    builder.setScaling(conf->keys, conf->values);
+  }
+
+  // Set specification for the grid tables to be used.
+  if (!conf->gridFiles.empty()) {
+    builder.setGridFiles(conf->gridFiles);
+  }
+
+  // Build the geometry from the DB blob
+
+  auto const& blob = iRecord.getTransientHandle(mayConsumeBlobToken_);
+  std::unique_ptr<std::vector<unsigned char> > tb = blob->getUncompressedBlob();
+
+  string sblob(tb->begin(), tb->end());
+  sblob.insert(
+      sblob.rfind("</DDDefinition>"),
+      "<MaterialSection label=\"materials.xml\"><ElementaryMaterial name=\"materials:Vacuum\" density=\"1e-13*mg/cm3\" "
+      "symbol=\" \" atomicWeight=\"1*g/mole\" atomicNumber=\"1\"/></MaterialSection>");
```

Comment:
this looks like a kludge.
Is this insert [still] needed?

Reply:
Yes. DD4Hep expects vaccuum to be defined, which was not the case in DDD. Current DB payloads do not therefore include its definition. Adding it here is simpler and safer than replacing all DB payloads.